athenet/sparsifying/utils/interval.py

```diff
@@ -0,0 +1,284 @@
+
+        # Input check below causes program interrupt if any _has_zero happened.
+        # return Interval(switch(self._has_zero(),
+        #                       T.constant(1)/T.constant(0),
+        #                       T.inv(self.upper)),
+        #                T.inv(self.lower))
+        return Interval(T.inv(self.upper), T.inv(self.lower))
+
+    def neg(self):
+        """For interval [a, b], returns interval [-b, -a]."""
+        return Interval(T.neg(self.upper), T.neg(self.lower))
+
+    def exp(self):
+        """Returns interval representing the exponential of the interval."""
+        return Interval(T.exp(self.lower), T.exp(self.upper))
+
+    def square(self):
+        """For interval I, returns I' such that for any x in I, I' contains
+        x*x and no other.
+
+        e.g [-1, 1].square() == [0, 1]"""
+        lsq = self.lower * self.lower
+        usq = self.upper * self.upper
+        u = T.maximum(lsq, usq)
+        l = T.switch(self._has_zero(), 0, T.minimum(lsq, usq))
+        return Interval(l, u)
+
+    def power(self, exponent):
+        """For interval i, returns i^exponent.
+
+        :exponent: Number to be passed as exponent to i^exponent.
+
+        Note: If interval contains some elements lower/equal to 0, exponent
+        should be integer."""
+        # If You want to understand what is happening here, make plot of
+        # f(x, y) = x^y domain. 'if's divide this domain with respect to
+        # monocity.
+        le = T.pow(self.lower, exponent)
+        ue = T.pow(self.upper, exponent)
+        l, u = None, None
+        if isinstance(exponent, (int, long)):
+            if exponent > 0:
+                if exponent % 2 == 0:
+                    l = T.switch(self._has_zero(), 0, T.minimum(le, ue))
+                    u = T.maximum(le, ue)
+                else:
+                    l = le
+                    u = ue
+            else:
+                if exponent % 2 == 0:
+                    l = T.minimum(le, ue)
+                    u = T.maximum(le, ue)
+                else:
+                    l = ue
+                    u = le
+        else:
+            # Assumes self.lower >= 0. Otherwise it is incorrectly defined.
+            # There is no check.
+            if exponent > 0:
+                l = le
+                u = ue
+            else:
+                l = ue
+                u = le
+        return Interval(l, u)
+
+    def dot(self, other):
+        """Dot product of Interval(self) vector and a number array (other)"""
+        # Requires project decision that could be better made after checking
+        # number of weights, edges and neurons in considered networks.
+        # TODO: Decide how to implement this. Decide whether not to consider
+        # batch in this implementation as it will be probably only used for
+        # batches.
+        raise NotImplementedError
+
+    def max(self, other):
+        """Returns interval such that for any (x, y) in (self, other),
```

Comment:
"for any (x, y) in (self, other)" - maybe say that (x, y) are pairs of intervals, it will be easier to understand. And you have to say at the beginning of class that "upper" and "lower" are Theano vectors.